reformatting input data for optimal space

tomyyyD opened the pull request:
Converts the user's format to a more dense format sorted by byte size. Stores in this dense format and then when the user retrieves data, converts back into the user's format.

tomyyyD committed:
added gitignore

tomyyyD committed:
added .pyc files to gitignore

tomyyyD committed:
Added dense Formatter

The formatter orders the fstring in order of byte size.

tomyyyD committed:
updates to make tests work now

tomyyyD committed:
Stores data in optimal format for space

takes user format and converts it into a more dense format for storage, then when retrieving data converts back to the users format.

tomyyyD committed:
removed unused imports for flake

thetazero commented on rexdb.py:
```diff
@@ -36,6 +38,61 @@ def col(self, i):
+        charList = {"c": 1, "?": 1, "h": 2, "i": 4, "f": 4}
+
+        # basic insertion sort to order the characters based on Byte size
+        for i in range(1, len(formatString)):
```
This function can be made much more simple and efficient using map and sort
```python3
# tie breaking by adding epsilon
charList = {"c": 1, "?": 0.1, "h": 2, "i": 4, "f": 4.1}
denseFormat = list(self._user_fstring)
denseFormat.sort(key=(lambda c: charList[c]), reverse=True)
return ''.join(denseFormat)
```

thetazero commented on rexdb.py:
```diff
@@ -36,6 +38,61 @@ def col(self, i):
                 data.append(line[i])
         return data[self._cursor:] + data[:self._cursor]
+
+    # formats for space c: char, i: int32, f: float, ?: bool, h: int16
```
Generally I'd like to split this packing code into its own class,
something like
```python3
packer = DensePacker('?ffi')
byte_data = packer.pack(True, 3.0, 4.0, 32)
print(packer.unpack(byte_data))
>>> (True, 3.0, 4.0, 32)
```

thetazero replied:
This should make it easier to write test for the functionality^

thetazero commented on rexdb.py:
```diff
@@ -36,6 +38,61 @@ def col(self, i):
+                                (3.2, "a", 5, 9.8) => (3.2, 5, 9.8, a) '''
+
+    def map_dense_user(self, data):
+        user_format = [*self._user_fstring]
```
list(self._user_fstring)

thetazero commented on rexdb.py:
```diff
@@ -36,6 +38,61 @@ def col(self, i):
+        # input is in dense_format
+        input = [d for d in data]
+        result = [0] * len(data)
+        for i in range(len(input)):
```
You should be able to do this in $O(n)$:
1. create a map from datatype --> empty lists
2. iterate through user_fstring and for each datatype append to the empty lists the index
3. iterate through the dense_fstring and pop an element from the list corresponding to the datatype, set result[i] to that index

thetazero replied:
After doing this computing the inverse should also be very easy, just store these two as a property of the dense_packer.

thetazero commented on rexdb.py:
```diff
@@ -36,6 +38,61 @@ def col(self, i):
+        return tuple(result)
+
+    # maps data in user_format to dense_format
+    def map_user_dense(self, data):
```
Same comments as map_dense_user.
Once you compute one direction, you should be able invert the map.

thetazero reviewed (changes requested):
Great start, I think this is a really good approach.

Once you refactor make sure to cover the `DensePacker` with unit tests.

tomyyyD committed:
Refactoring DensePacker

updating the functions to be simpler and more efficient and putting the packer in its own class.

tomyyyD committed:
Created the pack and unpack functions

data can be packed and unpacked based on a map that is created when DensePacker is initialized.

tomyyyD committed:
Flake8 Hates trailing spaces

thetazero commented on rexdb.py:
```diff
@@ -1,12 +1,76 @@
+        self.dense_fstring = self.make_format(fstring)
+        self.user_dense_map = self.create_pack_maps(self.user_fstring, self.dense_fstring)
+
+    '''
```
docstring into function

thetazero commented on rexdb.py:
```diff
@@ -36,6 +101,35 @@ def col(self, i):
                 data.append(line[i])
         return data[self._cursor:] + data[:self._cursor]
+
+    def map_dense_user(self, data: tuple) -> tuple:
```
dead code

tomyyyD committed:
update doc strings

thetazero committed:
Merge branch 'main' into dense_packer

tomyyyD committed:
updates to DensePacker functions

tomyyyD committed:
Merge branch 'dense_packer' of https://github.com/PyCubed-Mini/RexDB into dense_packer
new head

tomyyyD committed:
updates unit tests
